std-bits/base/src/main/java/org/enso/base/encoding/ReportingStreamDecoder.java

```diff
@@ -0,0 +1,316 @@
+   */
+  private CharBuffer outputBuffer = null;
+
+  /**
+   * The buffer keeping any input that has already been read but not decoded yet.
+   *
+   * <p>Between the calls to read, it satisfies the invariant that it is in 'reading' mode - to be
+   * able to write to it, it needs to be reallocated, compacted or flipped.
+   */
+  private ByteBuffer inputBuffer = null;
+
+  /**
+   * Indicates the amount of bytes consumed before the start of the current input buffer.
+   *
+   * <p>The input buffer is reset many times and so its position will only indicate the bytes that
+   * were consumed in some current iteration, this counter allows us to compute the overall amount
+   * of bytes.
+   */
+  private int inputBytesConsumedBeforeCurrentBuffer = 0;
+
+  /**
+   * We re-use the work array between calls to read, to avoid re-allocating it on each call. It is
+   * only re-allocated if it needs to be bigger than before.
+   */
+  private byte[] workArray = null;
+
+  /**
+   * A flag that is set once the end of input has been reached.
+   *
+   * <p>This informs us that no more input will be available from the input stream. There may still
+   * be some pending characters in the output buffer.
+   */
+  private boolean eof = false;
+
+  /**
+   * Specifies if the last {@code decoder.decode} call with the {@code endOfInput = true} argument
+   * has been made.
+   */
+  private boolean hadEofDecodeCall = false;
+
+  /**
+   * A list of positions containing encoding issues like malformed characters.
+   *
+   * <p>Used for reporting warnings.
+   */
+  List<Integer> encodingIssuePositions = new ArrayList<>();
+
+  @Override
+  public int read(char[] cbuf, int off, int len) throws IOException {
+    int readBytes = 0;
+
+    // First feed the pending characters that were already decoded.
+    if (outputBuffer != null && outputBuffer.hasRemaining()) {
+      int toTransfer = Math.min(len, outputBuffer.remaining());
+      outputBuffer.get(cbuf, off, toTransfer);
+      off += toTransfer;
+      len -= toTransfer;
+      readBytes += toTransfer;
+    }
+
+    // If the request is satisfied, we do not continue.
+    if (len <= 0) {
+      return readBytes;
+    }
+
+    // If we reached end of file, we won't be able to read any more data from the input. We also ran
+    // out of cached characters, so we indicate that there is no more input.
+    if (eof) {
+      // If the previous invocation of read set the EOF flag, it must have finished the decoding
+      // process and flushed the decoder, so the input buffer must have been consumed in whole.
+      assert !inputBuffer.hasRemaining();
+      return -1;
+    }
+
+    // At this point we ran out of cached characters, so we will read some more input to try to get
+    // new characters for the request.
+
+    prepareOutputBuffer(len);
+
+    int expectedInputSize = Math.max((int) (len / decoder.averageCharsPerByte()), 10);
```

Comment:
these three variables feel like they belong to readInputStreamToInputBuffer as not used else where